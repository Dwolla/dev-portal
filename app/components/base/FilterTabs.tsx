import React from "react";
import styled from "@emotion/styled";
import Select from "./select/Select";
import {
  GREY_1,
  GREY_2,
  GREY_4,
  GREY_6,
  GREY_8,
  WHITE_PRIMARY,
  ORANGE_PRIMARY,
} from "../colors";
import { POPPINS, ROBOTO } from "../typography";
import { breakUp, breakDown } from "../breakpoints";
import classnames from "../../modules/classnames";
import { slugify } from "../../modules/helpers";

// Styles
const TabsContainerStyle = styled.ul`
  margin: unset;
  padding: unset;
  list-style-type: none;
  justify-content: center;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 30px;

  @media (${breakDown("xs")}) {
    display: none;
  }

  &.sidebar {
    justify-content: left;
    display: block;
  }
`;

export const TabStyle = styled.li`
  height: 25px;
  color: ${GREY_4};
  font-family: ${POPPINS};
  font-size: 13px;
  font-weight: 500;
  line-height: 20px;
  justify-self: center;
  padding-bottom: 30px;
  cursor: pointer;

  :hover {
    color: ${GREY_2};
  }

  :focus {
    outline: none;
    color: ${WHITE_PRIMARY};
  }

  &.is-active {
    color: ${WHITE_PRIMARY};
    border-bottom: 3px solid ${ORANGE_PRIMARY};
  }

  &.sidebar {
    padding: 9px 0 11px 17px;
    height: 38px;
    min-width: 147px;
    color: ${GREY_6};
    background-color: ${WHITE_PRIMARY};
    font-family: ${ROBOTO};
    font-size: 14px;
    font-weight: unset;
    letter-spacing: 0;
    line-height: 19px;
    justify-self: left;
    box-sizing: border-box;
    border: 1px solid ${GREY_2};
    border-bottom: none;

    :last-child {
      border-bottom: 1px solid ${GREY_2};
    }

    :focus,
    :hover {
      color: ${GREY_8};
      padding-left: 16px;
      border-left: 2px solid ${GREY_6};
      background-color: ${GREY_1};
    }

    &.is-active {
      color: ${GREY_8};
      border-left: 2px solid ${ORANGE_PRIMARY};
      padding-left: 16px;
    }
  }
`;

// Tabs are replaced with Select dropdown on mobile
const SelectWrapper = styled.div`
  padding-bottom: 10px;

  @media (${breakUp("sm")}) {
    display: none;
  }

  &.sidebar {
    display: none;
  }
`;

type Tab = string | { value: string; label: string };
type Props = {
  tabs: Array<Tab>;
  filter: Tab | null;
  setFilter: Function;
  variant?: "default" | "sidebar";
};

export default function FilterTabs({
  tabs,
  filter,
  setFilter,
  variant,
}: Props) {
  const formattedTabs = tabs.map((tab) =>
    typeof tab === "string" ? { value: slugify(tab), label: tab } : tab
  );

  const formattedFilter =
    typeof filter === "string" || filter === null
      ? formattedTabs.find((tab) => tab.value === filter)
      : filter;

  function onKeydown(e, tab) {
    if (e.key === "Enter") setFilter(tab);
  }

  return (
    <>
      <TabsContainerStyle className={variant}>
        {formattedTabs.map((tab) => (
          <TabStyle
            tabIndex={0}
            className={classnames(variant, {
              "is-active":
                JSON.stringify(formattedFilter) === JSON.stringify(tab),
            })}
            key={tab.value}
            onClick={() => {
              setFilter(tab);
            }}
            onKeyDown={(keyPress) => onKeydown(keyPress, tab)}
          >
            {tab.label}
          </TabStyle>
        ))}
      </TabsContainerStyle>

      <SelectWrapper className={variant}>
        <Select
          options={formattedTabs}
          selectedValue={formattedFilter}
          setSelectedValue={setFilter}
          autoWidth
        />
      </SelectWrapper>
    </>
  );
}
