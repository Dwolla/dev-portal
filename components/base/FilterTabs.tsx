import React from "react";
import Dropdown from "react-dropdown";
import styled from "@emotion/styled";
import arrowDown from "../../public/images/arrow-down-icon.svg";
import arrowUp from "../../public/images/arrow-up-icon.svg";
import { GREY_4, WHITE_PRIMARY, PURPLE_DARK, ORANGE_PRIMARY } from "../colors";
import { POPPINS } from "../typography";
import { minWidth, maxWidth, BREAKPOINT_MOBILE } from "../breakpoints";
import classnames from "../../modules/classnames";

type Props = {
  tabs: string[];
  filter: string;
  setFilter: Function;
};

export default function FilterTabs({ tabs, filter, setFilter }: Props) {
  const dropdownOptions = [
    { value: "ALL", label: "ALL", className: "option" },
    ...tabs.map((item) => ({
      value: item,
      label: item,
      className: "option",
    })),
  ];
  const defaultOption = filter || dropdownOptions[0];

  return (
    <div>
      <TabsContainerStyle>
        <TabStyle
          className={classnames("tab", { "is-active": filter === null })}
          onClick={() => setFilter(null)}
        >
          ALL
        </TabStyle>

        {tabs.map((tab) => (
          <TabStyle
            className={classnames("tab", { "is-active": filter === tab })}
            key={tab}
            onClick={() => setFilter(tab)}
          >
            {tab}
          </TabStyle>
        ))}
      </TabsContainerStyle>

      <Drop>
        <Dropdown
          controlClassName="customControlClass"
          menuClassName="customMenuClass"
          placeholderClassName="customPlaceHolderClass"
          arrowOpen={<span className="arrowOpenClass" />}
          arrowClosed={<span className="arrowClosedClass" />}
          options={dropdownOptions}
          value={defaultOption}
          onChange={(e) => setFilter(e.label)}
        />
      </Drop>
    </div>
  );
}

// Styles
const TabsContainerStyle = styled.div`
  justify-content: center;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 30px;
  @media (${maxWidth(BREAKPOINT_MOBILE)}) {
    display: none;
  }
`;

export const TabStyle = styled.div`
  height: 25px;
  color: ${GREY_4};
  font-family: ${POPPINS};
  font-size: 13px;
  font-weight: 500;
  line-height: 20px;
  justify-self: center;
  padding-bottom: 30px;
  cursor: pointer;

  :hover {
    color: ${WHITE_PRIMARY};
  }

  &.is-active {
    color: ${WHITE_PRIMARY};
    border-bottom: 3px solid ${ORANGE_PRIMARY};
  }
`;

const Drop = styled.div`
  padding: 10px;
  @media (${minWidth(BREAKPOINT_MOBILE)}) {
    display: none;
  }

  /* Custom styles that are passed as props to react-dropdown component */
  .customControlClass {
    display: flex;
    justify-content: space-between;
  }
  .customControlClass,
  .customMenuClass {
    background-color: ${PURPLE_DARK};
    color: ${GREY_4};
    font-family: ${POPPINS};
    font-size: 13px;
    font-weight: 500;
    line-height: 20px;
    cursor: pointer;
  }

  .customPlaceHolderClass {
    color: ${WHITE_PRIMARY};
  }

  .option {
    color: inherit;
    margin-top: 5px;
    margin-bottom: 5px;
  }

  .option:hover {
    background-color: inherit;
    color: ${WHITE_PRIMARY};
  }

  .option.is-selected {
    background-color: inherit;
    color: ${WHITE_PRIMARY};
  }

  .arrowOpenClass,
  .arrowClosedClass {
    display: block;
    content: " ";
    padding: 10px;
    width: auto;
  }

  .arrowOpenClass {
    background: url(${arrowUp}) no-repeat center;
  }

  .arrowClosedClass {
    background: url(${arrowDown}) no-repeat center;
  }
`;
