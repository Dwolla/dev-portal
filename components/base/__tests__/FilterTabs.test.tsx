import renderer from "react-test-renderer";
import FilterTabs, { TabStyle } from "../FilterTabs";

const TABS = ["GET BUILDING", "PREREQUISITES", "FEATURES"];

test("FilterTabs", () => {
  const setState = jest.fn();
  const tree = renderer.create(
    <FilterTabs tabs={TABS} filter={null} setFilter={setState} />
  );
  tree.root.findAllByType(TabStyle).map((t) => t.props.onClick());

  expect(setState.mock.calls.map((c) => c)).toEqual(
    [null, ...TABS].map((t) => [t])
  );
  expect(tree.toJSON()).toMatchSnapshot();
});
