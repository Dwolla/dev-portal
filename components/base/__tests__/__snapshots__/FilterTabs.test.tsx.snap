// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterTabs 1`] = `
<div>
  <div
    className="css-18c9go6-TabsContainerStyle efq5uj30"
  >
    <div
      className="tab is-active css-rdl1v4-TabStyle efq5uj31"
      onClick={[Function]}
    >
      ALL
    </div>
    <div
      className="tab css-rdl1v4-TabStyle efq5uj31"
      onClick={[Function]}
    >
      GET BUILDING
    </div>
    <div
      className="tab css-rdl1v4-TabStyle efq5uj31"
      onClick={[Function]}
    >
      PREREQUISITES
    </div>
    <div
      className="tab css-rdl1v4-TabStyle efq5uj31"
      onClick={[Function]}
    >
      FEATURES
    </div>
  </div>
  <div
    className="css-n6mzzn-SelectWrapper efq5uj32"
  >
    <div
      className=" css-2b097c-container"
      onKeyDown={[Function]}
    >
      <div
        className=" css-1xxipdn-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className=" css-86rg1c-ValueContainer"
        >
          <div
            className=" css-1me8bqy-singleValue"
          >
            ALL
          </div>
          <input
            aria-autocomplete="list"
            className="css-62g3xt-dummyInput"
            disabled={false}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            tabIndex="0"
            value=""
          />
        </div>
        <div
          className=" css-1hb7zxy-IndicatorsContainer"
        >
          <span
            className=" css-43ykx9-indicatorSeparator"
          />
          <div
            aria-hidden="true"
            className=" css-unjupu-indicatorContainer"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <img
              alt=""
              className=" css-vfym8x-StyledDefaultDropdownIcon e1oqaitj0"
              src="test-image-stub"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
